Linkification testing

macchiati opened the pull request:
Some preliminary code to check the linkification algorithm and generate test cases.

macchiati commented:
TODO:

1. Update data file & test data generator to match rev 1 draft 5.
2. Add newer test data from ICANN.
3. Change the data file folder to /Public/<version>/linkification/
4. Change the filename SerializationTest.txt to FormattingTest.txt
5. Create a unicodetools PR with the code, get it reviewed & merged.

markusicu commented on unicodetools/data/linkification/dev/LinkEmail.txt:
```diff
@@ -0,0 +1,1298 @@
+# Format
+#
+# Field 0: code point range
+# Field 1: binary value
```
I guess this works, but it looks a bit odd. The second column is basically useless.

This must be the first time that we have a data file dedicated to one single **binary** property.

When we list binary properties in multi-property data files, we don't have a value column. By analogy, we only really need a single code point column here.

markusicu commented on unicodetools/data/linkification/dev/LinkEmail.txt:
```diff
@@ -0,0 +1,1298 @@
+#  All code points not explicitly listed for Link_Email
+#  have the value No.
+#
+# @missing: 0000..10FFFF; No
```
same as above -- we haven't written `@missing` lines for binary properties; they just all default to `No`

of course the `@missing` line, when present in a data file, should have the same syntax as the value lines in the same file. so if we explicitly print `Yes` then i guess we should have this line as well. but i think it just shows even more how weird it is. i think we should omit the `Yes` values and also omit the `@missing` line.

markusicu commented on unicodetools/data/linkification/dev/LinkTerm.txt:
```diff
@@ -0,0 +1,2156 @@
+# Field 1: a Link_Term value
+#          For more information, see https://www.unicode.org/reports/tr58/#property-data. 
+#
+# For the purpose of detection and formatting operations, the property Link_Term is defined as
```
We have been writing "For the purpose of regular expressions" as a hook for documenting names/types/values of non-UCD properties.

markusicu commented on unicodetools/data/linkification/dev/LinkDetectionTest.txt:
```diff
@@ -0,0 +1,434 @@
+See ..john.doe@example.com
+
+#Quoted local-parts (not in the base algorithm).
+See "john\ ⸠doe@example.com⸡"
```
in formal email syntax, the closing quote would be right before the at sign, correct?

markusicu commented on unicodetools/data/linkification/dev/LinkFormattingTest.txt:
```diff
@@ -0,0 +1,126 @@
+# Field 4: Result — with minimal escaping
+#
+# Empty lines, and lines starting with # are ignored.
+# Spaces around the semicolons are ignored.
```
Please swap this line with the next one.

markusicu reviewed (changes requested):
Review of only the data files so far.

macchiati commented:
I think I got all of them except for the @example.com; that needs a code
fix that I'll get to later today.

On Mon, Dec 15, 2025 at 8:17 PM Markus Scherer ***@***.***>
wrote:

> ***@***.**** requested changes on this pull request.
>
> Review of only the data files so far.
> ------------------------------
>
> In unicodetools/data/linkification/dev/LinkBracket.txt
> <https://github.com/unicode-org/unicodetools/pull/961#discussion_r2621620682>
> :
>
> > +# Property:	Link_Bracket
> +# Format
> +#
> +# Field 0: code point
> +# Field 1: code point
> +#          For more information, see https://www.unicode.org/reports/tr58/#property-data.
> +#
> +# For the purpose of link detection and formatting operations, the property Link_Bracket is defined as
> +# a string property whose value is either a single code point or is undefined.
> +#
> +# The short name of the property is the same as its long name.
> +#
> +#  All code points not explicitly listed for Link_Bracket
> +#  have the value undefined.
> +#
> +# @missing: 0000..10FFFF; undefined
>
> I don't think we have ever had an undefined missing value.
> Let's make it the same as for Bidi_Paired_Bracket and lots of other
> properties:
> ⬇️ Suggested change
>
> -# @missing: 0000..10FFFF; undefined
> +# @missing: 0000..10FFFF; <none>
>
> done
(Thanks, you can tell I haven't done this in a while)


>
> ------------------------------
>
> In unicodetools/data/linkification/dev/LinkBracket.txt
> <https://github.com/unicode-org/unicodetools/pull/961#discussion_r2621625007>
> :
>
> > +# ================================================
> +#
> +# Property:	Link_Bracket
> +# Format
> +#
> +# Field 0: code point
> +# Field 1: code point
> +#          For more information, see https://www.unicode.org/reports/tr58/#property-data.
> +#
> +# For the purpose of link detection and formatting operations, the property Link_Bracket is defined as
> +# a string property whose value is either a single code point or is undefined.
> +#
> +# The short name of the property is the same as its long name.
> +#
> +#  All code points not explicitly listed for Link_Bracket
> +#  have the value undefined.
>
> ⬇️ Suggested change
>
> -#  have the value undefined.
> +#  have the value <none>.
>
> see below
> ------------------------------
>
> In unicodetools/data/linkification/dev/LinkBracket.txt
> <https://github.com/unicode-org/unicodetools/pull/961#discussion_r2621626836>
> :
>
> > +#          For more information, see https://www.unicode.org/reports/tr58/#property-data.
> +#
> +# For the purpose of link detection and formatting operations, the property Link_Bracket is defined as
> +# a string property whose value is either a single code point or is undefined.
> +#
> +# The short name of the property is the same as its long name.
> +#
> +#  All code points not explicitly listed for Link_Bracket
> +#  have the value undefined.
> +#
> +# @missing: 0000..10FFFF; undefined
> +#
> +# ================================================
> +
> +0029          ; 0028 #1.1     () ⇒ ()       RIGHT PARENTHESIS
> +003E          ; 003C #1.1     (&gt; ⇒ &lt;)       GREATER-THAN SIGN
>
> escaping < and > in this context seems weird
>
Ah, I'd noticed that but forgot to fix it. It uses a transliterator to make
sure values are kosher, but I had the wrong one.
fixed


> ------------------------------
>
> In unicodetools/data/linkification/dev/LinkEmail.txt
> <https://github.com/unicode-org/unicodetools/pull/961#discussion_r2621637227>
> :
>
> > @@ -0,0 +1,1298 @@
> +# LinkEmail.txt
> +# Date: 2025-12-14, 06:39:23 GMT
> +# © 2025 Unicode®, Inc.
> +# Unicode and the Unicode Logo are registered trademarks of Unicode, Inc. in the U.S. and other countries.
> +# For terms of use and license, see https://www.unicode.org/terms_of_use.html
> +#
> +# The usage and stability of these values is covered in https://www.unicode.org/reports/tr58/
> +#
> +# ================================================
> +#
> +# Property:	Link_Email
> +# Format
> +#
> +# Field 0: code point range
> +# Field 1: binary value
>
> I guess this works, but it looks a bit odd. The second column is basically
> useless.
>
> This must be the first time that we have a data file dedicated to one
> single *binary* property.
>
> When we list binary properties in multi-property data files, we don't have
> a value column. By analogy, we only really need a single code point column
> here.
> ------------------------------
>
> In unicodetools/data/linkification/dev/LinkEmail.txt
> <https://github.com/unicode-org/unicodetools/pull/961#discussion_r2621638283>
> :
>
> > +# Property:	Link_Email
> +# Format
> +#
> +# Field 0: code point range
> +# Field 1: binary value
> +#          For more information, see https://www.unicode.org/reports/tr58/#property-data.
> +#
> +# For the purpose of link detection and formatting operations, the property Link_Email is defined as
> +# a binary property.
> +#
> +# The short name of the property is the same as its long name.
> +#
> +#  All code points not explicitly listed for Link_Email
> +#  have the value No.
> +#
> +# @missing: 0000..10FFFF; No
>
> same as above -- we haven't written @missing lines for binary properties;
> they just all default to No
>
> of course the @missing line, when present in a data file, should have the
> same syntax as the value lines in the same file. so if we explicitly print
> Yes then i guess we should have this line as well. but i think it just
> shows even more how weird it is. i think we should omit the Yes values
> and also omit the @missing line.
>
done

------------------------------
>
> In unicodetools/data/linkification/dev/LinkTerm.txt
> <https://github.com/unicode-org/unicodetools/pull/961#discussion_r2621644897>
> :
>
> > +# Unicode and the Unicode Logo are registered trademarks of Unicode, Inc. in the U.S. and other countries.
> +# For terms of use and license, see https://www.unicode.org/terms_of_use.html
> +#
> +# The usage and stability of these values is covered in https://www.unicode.org/reports/tr58/
> +#
> +# ================================================
> +#
> +# Property:	Link_Term
> +# Format
> +#
> +# Field 0: code point range
> +# Field 1: a Link_Term value
> +#          For more information, see https://www.unicode.org/reports/tr58/#property-data.
> +#
> +# For the purpose of detection and formatting operations, the property Link_Term is defined as
> +# mapping each code point to a set of enumerated values.
>
> ⬇️ Suggested change
>
> -# mapping each code point to a set of enumerated values.
> +# an enumerated property of code points.
>
> like in
> https://www.unicode.org/Public/17.0.0/security/IdentifierStatus.txt
>
done

> ------------------------------
>
> In unicodetools/data/linkification/dev/LinkTerm.txt
> <https://github.com/unicode-org/unicodetools/pull/961#discussion_r2621648015>
> :
>
> > +# © 2025 Unicode®, Inc.
> +# Unicode and the Unicode Logo are registered trademarks of Unicode, Inc. in the U.S. and other countries.
> +# For terms of use and license, see https://www.unicode.org/terms_of_use.html
> +#
> +# The usage and stability of these values is covered in https://www.unicode.org/reports/tr58/
> +#
> +# ================================================
> +#
> +# Property:	Link_Term
> +# Format
> +#
> +# Field 0: code point range
> +# Field 1: a Link_Term value
> +#          For more information, see https://www.unicode.org/reports/tr58/#property-data.
> +#
> +# For the purpose of detection and formatting operations, the property Link_Term is defined as
>
> We have been writing "For the purpose of regular expressions" as a hook
> for documenting names/types/values of non-UCD properties.
>
Ah, didn't know the right incantation
done


> ------------------------------
>
> In unicodetools/data/linkification/dev/LinkBracket.txt
> <https://github.com/unicode-org/unicodetools/pull/961#discussion_r2621648359>
> :
>
> > +# © 2025 Unicode®, Inc.
> +# Unicode and the Unicode Logo are registered trademarks of Unicode, Inc. in the U.S. and other countries.
> +# For terms of use and license, see https://www.unicode.org/terms_of_use.html
> +#
> +# The usage and stability of these values is covered in https://www.unicode.org/reports/tr58/
> +#
> +# ================================================
> +#
> +# Property:	Link_Bracket
> +# Format
> +#
> +# Field 0: code point
> +# Field 1: code point
> +#          For more information, see https://www.unicode.org/reports/tr58/#property-data.
> +#
> +# For the purpose of link detection and formatting operations, the property Link_Bracket is defined as
>
> We have been writing "For the purpose of regular expressions" as a hook
> for documenting names/types/values of non-UCD properties.
>
done

> ------------------------------
>
> In unicodetools/data/linkification/dev/LinkEmail.txt
> <https://github.com/unicode-org/unicodetools/pull/961#discussion_r2621648667>
> :
>
> > +# © 2025 Unicode®, Inc.
> +# Unicode and the Unicode Logo are registered trademarks of Unicode, Inc. in the U.S. and other countries.
> +# For terms of use and license, see https://www.unicode.org/terms_of_use.html
> +#
> +# The usage and stability of these values is covered in https://www.unicode.org/reports/tr58/
> +#
> +# ================================================
> +#
> +# Property:	Link_Email
> +# Format
> +#
> +# Field 0: code point range
> +# Field 1: binary value
> +#          For more information, see https://www.unicode.org/reports/tr58/#property-data.
> +#
> +# For the purpose of link detection and formatting operations, the property Link_Email is defined as
>
> We have been writing "For the purpose of regular expressions" as a hook
> for documenting names/types/values of non-UCD properties.
>
done


> ------------------------------
>
> In unicodetools/data/linkification/dev/LinkDetectionTest.txt
> <https://github.com/unicode-org/unicodetools/pull/961#discussion_r2621664135>
> :
>
> > @@ -0,0 +1,434 @@
> +# LinkDetectionTest.txt
> +# Date: 2025-12-16, 01:20:14 GMT
> +# © 2025 Unicode®, Inc.
> +# Unicode and the Unicode Logo are registered trademarks of Unicode, Inc. in the U.S. and other countries.
> +# For terms of use and license, see https://www.unicode.org/terms_of_use.html
> +#
> +# The usage and stability of these values is covered in https://www.unicode.org/reports/tr58/
> +#
> +# ================================================
> +#
> +# Format:
> +#   Each line contains zero or more marked links, such as ⸠abc.com⸡
> +#
> +# Operation:#   For each line.
>
> ⬇️ Suggested change
>
> -# Operation:#   For each line.
> +# Operation:
> +#   For each line.
>
> done

> ------------------------------
>
> In unicodetools/data/linkification/dev/LinkDetectionTest.txt
> <https://github.com/unicode-org/unicodetools/pull/961#discussion_r2621669023>
> :
>
> > +# Misc. test cases
> +
> +# Implementations may differ as to whether they linkify text where the TLD is invalid (not listed in  IANA's tlds-alpha-by-domain.txt)
> +# So this is not specifically tested for
> +# a test.invalid b
> +
> +# Sample uppercase
> +See ⸠TEST.COM⸡ on…
> +See ⸠FOO.VERMÖGEN.com <http://foo.xn--vermgen-d1a.com>⸡ on…
> +
> +# Illegal domain names
> +See http://.foo.example.com/αβγ <http://foo.example.com/%CE%B1%CE%B2%CE%B3> on…
> +See http://foo..example.com/αβγ <http://example.com/%CE%B1%CE%B2%CE%B3> on…
> +See http://-foo.example-.com. on…
> +
> +# Legal but unusual. Because we might be at the end of a sentence, we don't include the . unless followed by [/$#]
>
> What does the $ stand for?
>
Typo, changed to followed by a Path, Query, or Fragment

> ------------------------------
>
> In unicodetools/data/linkification/dev/LinkDetectionTest.txt
> <https://github.com/unicode-org/unicodetools/pull/961#discussion_r2621679306>
> :
>
> > +See ***@***.*** on…
> +
> +#Stop backing up when a space is hit
> +See ***@***.***⸡
> +
> +#Include the medial dot
> +See ***@***.***⸡.
> +
> +#Handle non-ASCII
> +See ***@***.***⸡
> +
> +#No valid domain name
> +See @example.😎
> +
> +# No local-part
> +See ***@***.***⸡
>
> should this really linkify?
>
That's a mistake in my code; it shouldn't. Will fix


> ------------------------------
>
> In unicodetools/data/linkification/dev/LinkDetectionTest.txt
> <https://github.com/unicode-org/unicodetools/pull/961#discussion_r2621680678>
> :
>
> > +#Handle non-ASCII
> +See ***@***.***⸡
> +
> +#No valid domain name
> +See @example.😎
> +
> +# No local-part
> +See ***@***.***⸡
> +
> +# No valid local-part
> +See ***@***.***
> +See .***@***.***
> +See ..***@***.***
> +
> +#Quoted local-parts (not in the base algorithm).
> +See "john\ ***@***.***⸡"
>
> in formal email syntax, the closing quote would be right before the at
> sign, correct?
>
Yes. Fixed the sample, but still have to fix the @example problem.

------------------------------
>
> In unicodetools/data/linkification/dev/LinkFormattingTest.txt
> <https://github.com/unicode-org/unicodetools/pull/961#discussion_r2621686042>
> :
>
> > +# Unicode and the Unicode Logo are registered trademarks of Unicode, Inc. in the U.S. and other countries.
> +# For terms of use and license, see https://www.unicode.org/terms_of_use.html
> +#
> +# The usage and stability of these values is covered in https://www.unicode.org/reports/tr58/
> +#
> +# ================================================
> +#
> +# Format: Each line has the following fields, separated by semicolons.
> +# Field 0: Scheme/host
> +# Field 1: Path
> +# Field 2: Query
> +# Field 3: Fragment
> +# Field 4: Result — with minimal escaping
> +#
> +# Empty lines, and lines starting with # are ignored.
> +# Spaces around the semicolons are ignored.
>
> Please swap this line with the next one.
>
done

> ------------------------------
>
> In unicodetools/data/linkification/dev/LinkFormattingTest.txt
> <https://github.com/unicode-org/unicodetools/pull/961#discussion_r2621690037>
> :
>
> > +# The usage and stability of these values is covered in https://www.unicode.org/reports/tr58/
> +#
> +# ================================================
> +#
> +# Format: Each line has the following fields, separated by semicolons.
> +# Field 0: Scheme/host
> +# Field 1: Path
> +# Field 2: Query
> +# Field 3: Fragment
> +# Field 4: Result — with minimal escaping
> +#
> +# Empty lines, and lines starting with # are ignored.
> +# Spaces around the semicolons are ignored.
> +# Otherwise # is treated like any other character.
> +#
> +# The Path, Query, and Fragment will contain backslash escapes when characters would otherwise be
>
> I have questions in the "Notes" doc about the purpose and behavior of the
> backslash syntax. You didn't respond there. I think we should drop this.
>
We can't. But let's discuss in the meeting.


> ------------------------------
>
> In unicodetools/data/linkification/dev/LinkFormattingTest.txt
> <https://github.com/unicode-org/unicodetools/pull/961#discussion_r2621698408>
> :
>
> > +https:// ;	example.com ;	 ;	α%β%3D=γ%δ%3D ;	 ;	https://example.com?α%β%253D=γ%δ%253D
> +
> +# Wikipedia test cases
> +
> +https:// ;	ru.wikinews.org ;	/wiki/Категория:Вселенная ;	 ;	 ;	https://ru.wikinews.org/wiki/Категория:Вселенная
> +https:// ;	av.wikipedia.org ;	/wiki/Ракь_(планета) ;	 ;	 ;	https://av.wikipedia.org/wiki/Ракь_(планета)
> +https:// ;	bo.wikipedia.org ;	/wiki/སའི་གོ་ལ། ;	 ;	 ;	https://bo.wikipedia.org/wiki/སའི་གོ་ལ%E0%BC%8D
> +https:// ;	fiu-vro.wikipedia.org ;	/wiki/Maa_(hod'otäht) ;	 ;	 ;	https://fiu-vro.wikipedia.org/wiki/Maa_(hod'otäht)
> +https:// ;	ty.wikipedia.org ;	/wiki/’Afirita ;	 ;	 ;	https://ty.wikipedia.org/wiki/’Afirita
> +https:// ;	ab.wikipedia.org ;	/wiki/Вашингтон,_Џьорџь ;	 ;	 ;	https://ab.wikipedia.org/wiki/Вашингтон,_Џьорџь
> +https:// ;	mni.wikipedia.org ;	/wiki/ꯅ꯭ꯌꯨ_ꯌꯣꯔ꯭ꯛ_ꯁꯤꯇꯤꯒꯤ_ꯌꯨ.ꯑꯦꯁ. ;	 ;	 ;	https://mni.wikipedia.org/wiki/ꯅ꯭ꯌꯨ_ꯌꯣꯔ꯭ꯛ_ꯁꯤꯇꯤꯒꯤ_ꯌꯨ.ꯑꯦꯁ%2E
> +https:// ;	azb.wikipedia.org ;	/wiki/واشینقتن،_دی.سی. ;	 ;	 ;	https://azb.wikipedia.org/wiki/واشینقتن،_دی.سی%2E
> +https:// ;	mad.wikipedia.org ;	/wiki/Tasè’ ;	 ;	 ;	https://mad.wikipedia.org/wiki/Tasè%E2%80%99 <https://mad.wikipedia.org/wiki/Tas%C3%A8%E2%80%99>
> +https:// ;	wuu.wikipedia.org ;	/wiki/聖保羅（巴西） ;	 ;	 ;	https://wuu.wikipedia.org/wiki/聖保羅（巴西）
> +https:// ;	vep.wikipedia.org ;	/wiki/Brüssel' ;	 ;	 ;	https://vep.wikipedia.org/wiki/Brüssel%27 <https://vep.wikipedia.org/wiki/Br%C3%BCssel%27>
> +https:// ;	tw.wikipedia.org ;	/wiki/Wiase_Nyinaa_Wɛbsaet_(_World_Wide_Web;_WWW_) ;	 ;	 ;	https://tw.wikipedia.org/wiki/Wiase_Nyinaa_Wɛbsaet_(_World_Wide_Web;_WWW_)
>
> you can't have a literal ; in a field in a semicolon-separated file --
> see the "Notes" doc
>
Ok, I'll filter those out.

> —
> Reply to this email directly, view it on GitHub
> <https://github.com/unicode-org/unicodetools/pull/961#pullrequestreview-3581066076>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACJLEMCCVXUQUSU6YGRJBAL4B6BW3AVCNFSM6AAAAACICKNJ5KVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZTKOBRGA3DMMBXGY>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

markusicu commented:
> Mark: Please respond to GitHub comments on GitHub, not via email -- email replies show up in the stream, but are separate from what they respond to, and are a pain to track.

FYI -- I resolved my comments that you took care of, keeping the unresolved visible.
I will continue reviewing other files today.

macchiati commented:
Ok, and thanks for the detailed review. I had a brain-hiccough; should have
responded in github.

On Tue, Dec 16, 2025 at 10:21 AM Markus Scherer ***@***.***>
wrote:

> *markusicu* left a comment (unicode-org/unicodetools#961)
> <https://github.com/unicode-org/unicodetools/pull/961#issuecomment-3661835887>
>
> Mark: Please respond to GitHub comments on GitHub, not via email -- email
> replies show up in the stream, but are separate from what they respond to,
> and are a pain to track.
>
> FYI -- I resolved my comments that you took care of, keeping the
> unresolved visible.
> I will continue reviewing other files today.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/unicode-org/unicodetools/pull/961#issuecomment-3661835887>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACJLEMBO4UBRL4RNW5GDDRD4CBEUNAVCNFSM6AAAAACICKNJ5KVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTMNRRHAZTKOBYG4>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

markusicu commented on unicodetools/data/linkification/dev/LinkFormattingTest.txt:
```diff
@@ -0,0 +1,125 @@
+# ================================================
+
+
+# Selected test cases
```
This file has no test cases for formatting emails. Do we no longer need that because we no longer handle quoted email local parts where escaping is even possible?

markusicu commented on unicodetools/src/main/resources/org/unicode/tools/testUrlsStats.txt:
What is the purpose of this file?

macchiati replied:
It is just a summary of the number of URLs in each locale from the large file. I can delete that.

macchiati replied:
deleted

markusicu commented on unicodetools/src/test/java/org/unicode/propstest/CompareVersionedProps.java:
```diff
@@ -24,6 +25,9 @@
 public class CompareVersionedProps {
+    private static IndexUnicodeProperties IUP = IndexUnicodeProperties.make();
+
+    private static final boolean NO_EMOJI = true;
```
Is this necessary because of the current not-fully-Unicode-18 state of data?
Will this affect other tools and tests?
When do we need to flip this to false?

markusicu commented on unicodetools/src/main/java/org/unicode/utilities/LinkUtilities.java:
```diff
@@ -0,0 +1,1146 @@
+                        new UnicodeSet(
+                                        uset,
+                                        parsePosition,
+                                        VersionedSymbolTable.frozenAt(VersionInfo.UNICODE_17_0))
```
latest version

also, it's error-prone to have multiple places in a file that set a Unicode version

macchiati replied:
Will combine, with comment.

markusicu commented on unicodetools/src/main/java/org/unicode/utilities/LinkUtilities.java:
```diff
@@ -0,0 +1,1146 @@
+            new UnicodeSet(idnValid)
+                    .addAll(idnMapped)
+                    .removeAll(new UnicodeSet("[:Block=Basic_Latin:]"))
+                    .addAll(new UnicodeSet("[-a-zA-Z0-9..]"))
```
why two dots?

macchiati replied:
Fixed this. Tho not currently used by the tests (because we don't focus on the domain name) the dots ".．。" are treated as identical in toUTS46n.

macchiati replied:
Didn't resolve this yet to give you a chance to see the change.

markusicu commented on unicodetools/src/main/java/org/unicode/utilities/LinkUtilities.java:
```diff
@@ -0,0 +1,1146 @@
+                    .addAll(new UnicodeSet("[-a-zA-Z0-9..]"))
+                    .freeze();
+    public static final UnicodeSet validHostNoDot =
+            new UnicodeSet(validHost).remove('.').remove('.').freeze();
```
again why two dots?

markusicu commented on unicodetools/data/linkification/dev/LinkDetectionTest.txt:
```diff
@@ -0,0 +1,432 @@
+See ..john.doe@example.com
+
+#Quoted local-part, so not linked (in the base algorithm)
+See "john\ doe"@example.com
```
In your latest spec draft, you write “Implementations are free to extend this algorithm to support such quoted email local-part formats.”

If an implementation does that, then it will fail this test case.
I suggest we remove it.

macchiati commented on unicodetools/data/linkification/dev/LinkFormattingTest.txt:
```diff
@@ -0,0 +1,126 @@
+# Spaces around the semicolons are ignored.
+# Otherwise # is treated like any other character.
+#
+# The Path, Query, and Fragment will contain backslash escapes when characters would otherwise be 
```
For the tests, there needs to be some way on the input side to distinguish (eg in a query) between a "=" that is a literal, and a "=" that is in a key-value pair. That is:

foo.com?a=b=c
representing 
{domain="foo.com" query={{key="a" value="b=c"}}}

We need to be able have something that is readable, and to:
-  distinguish the 2nd = as being a literal, not a syntax character, OR
-  distinguish the 1st = as being a syntax character, and not a literal

I tried %xx for the literals, but that completely obfuscated the input string.

The characters we need to worry about are (depending on the context):

, ? / & # =

I thought of using non-ascii characters for either the literals or the syntax characters. Not sure whether that would work, however.

macchiati commented on unicodetools/src/main/resources/org/unicode/tools/LinkDetectionTestSourceICANN.txt:
```diff
@@ -0,0 +1,145 @@
+## Original file is from ICANN
```
They donated this under CLA. I can ask them if they want a copyright ack.

macchiati commented on unicodetools/src/main/resources/org/unicode/tools/tlds-alpha-by-domain.txt:
I disagree. The tests don't depend on it any more, but it is a more real-world check on what we are doing.

macchiati commented on unicodetools/src/main/java/org/unicode/utilities/LinkUtilities.java:
```diff
@@ -0,0 +1,1146 @@
+        HOST('\u0000', "[/?#]", "[]", "[]"),
+        PATH('/', "[?#]", "[/]", "[]"),
+        QUERY('?', "[#]", "[=\\&]", "[+]"),
+        FRAGMENT('#', "[]", "[]", "[\\{:~\\}]");
```
Whoops, fixed.

macchiati commented on unicodetools/src/main/java/org/unicode/utilities/LinkUtilities.java:
```diff
@@ -0,0 +1,1146 @@
+public class LinkUtilities {
+    // allow changing UnicodeSet to use the current IndexUnicodeProperties
+    public static final IndexUnicodeProperties IUP =
+            IndexUnicodeProperties.make(VersionInfo.UNICODE_17_0);
```
Agreed. That's why I set it to a hard-coded version for now.

macchiati replied:
I agree that we should _also_ generate the 18.0 data.

macchiati commented on unicodetools/src/main/java/org/unicode/utilities/LinkUtilities.java:
```diff
@@ -0,0 +1,1146 @@
+            IndexUnicodeProperties.make(VersionInfo.UNICODE_17_0);
+
+    public static final boolean DEBUG = false;
+    public static final boolean USE_CLDR = false;
```
Sure, can remove.

macchiati commented on unicodetools/src/main/java/org/unicode/utilities/LinkUtilities.java:
```diff
@@ -0,0 +1,1146 @@
+        final int initiator;
+        final UnicodeSet terminators;
+        final UnicodeSet clearStack;
+        final UnicodeSet extraQuoted;
```
Good catch.
What is it for? A literal + within a query needs to be quoted; otherwise it can be interpreted as a space. I'll add a note to the spec about that.

macchiati commented:
Added a note about the + in queries in minimal escaping:
https://unicode-org.github.io/unicode-reports/pull/247/tr58/tr58.html#minimal-escaping-algorithm

markusicu commented on target/spotless-index:
did you add this accidentally?

markusicu reviewed (commented):
One more comment for now, and there is at least one of the previous comment without discussion. I am also going over your discussion doc.
